README.md
# NAME

Markdown::Pod - Convert Markdown to POD

# VERSION

version 0.005

# SYNOPSIS

    use Markdown::Pod;
    
    my $m2p = Markdown::Pod->new;
    my $pod = $m2p->markdown_to_pod(
        markdown => $markdown,
    );

# DESCRIPTION

This module parses Markdown text and return POD text.
It uses [Markdent](https://metacpan.org/pod/Markdent) module to parse Markdown.
Due to POD doesn't support blockquoted HTML tag,
so quoted text of Markdown will not be handled properly.
Quoted text will be converted to POD verbatim section.

# ATTRIBUTES

## markdown

markdown text

## encoding

encoding to use. Available type of encoding is same as [Encode](https://metacpan.org/pod/Encode) module.

# METHODS

## new

create Markdown::Pod object

## markdown\_to\_pod

convert markdown text to POD text

# SEE ALSO

- [Markdent](https://metacpan.org/pod/Markdent)
- [Pod::Markdown](https://metacpan.org/pod/Pod::Markdown)
- [Text::MultiMarkdown](https://metacpan.org/pod/Text::MultiMarkdown), [Text::Markdown](https://metacpan.org/pod/Text::Markdown)

# AUTHOR

Keedi Kim - 김도형 <keedi@cpan.org>

# CONTRIBUTORS

- Abigail (ABIGAIL)
- Jason McIntosh (JMAC)

# COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Keedi Kim.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for {{ $dist->name }}

{{ $NEXT }}

0.008     2021-04-11 21:35:57+09:00 Asia/Seoul
    - Remove double spaces if attributes are empty
    - [GitHub #13] Fix boolean HTML attribute handling bug (POTATOGIM)
      - Update module dependency: Markdent 0.40
      - Reported-by: SREZIC, DROLSKY

0.007     2018-02-24 17:58:42+09:00 Asia/Seoul
    - [GitHub #7] Fix nested list converting bug (ZMUGHAL)
      - Add extra newline when starting a nested list
      - Reported-by: JMAC, olafhering
    - [GitHub #8] Markdent v0.27 uses Specio for types (ZMUGHAL)
      - Fixes an installation error that occurs when using the version
        v0.27 or later of Markdent using Specio
      - Reported-by: markwilkinson, rsrpsinr

0.006     2015-02-22 17:41:28+09:00 Asia/Seoul
    - [GitHub #6] Use a stack to handle styles inside other styles (ZMUGHAL)
    - [GitHub #5] Smarter use of code formatting (ZMUGHAL)
    - [GitHub #4] Formatting fixes (ASPEER)
        - Adding --dialect option to script command line
        - Adding missing methods (e.g. auto_link, horizontal rule)
          to Handler.pm
        - Adding simple table formatting to Handler.pm
        - Changing block-quote formatting so it produces output
          compatible with Pod::Simple::HTML

0.005     2013-12-15 17:50:43 Asia/Seoul
    - remove File::Slurp module dependency
    - [GitHub #3] insisting on Markdent v0.23 or later,
      in order to properly support lazy list numbering. (JMAC)

0.004     2013-12-12 16:35:19 Asia/Seoul
    - [GitHub #2] add handler support for HTML entities (JMAC)

0.003     2012-07-26 21:04:23 Asia/Seoul
    - handle Markdent 0.22's line_break event
      - separate paragraph if line_break is occurred
    - [RT #77887] parse embed html
      - display html using POD's =begin and =end tag pair
    - [RT #77889] display list's bullet
      - Markdent 0.21 supports bullet attribute for list

0.002     2012-06-18 18:06:55 Asia/Seoul
    - add real world markdown converting test
      - take markdown from http://advent.perl.kr/2011
    - do not check attr param for end_html_tag
    - [RT #75458] apply abigail's patch
      - emphasis tag will be match with POD's 'I<>'
      - strong tag will be match with POD's 'B<>'

0.001     2011-12-10 04:15:49 Asia/Seoul
    First version, released on unsuspecting world.


t/lib/Markdown/Pod/Test.pm
use strict;
use warnings;

package Markdown::Pod::Test;

use Exporter 5.57 qw(import);
use Encode        qw(decode_utf8);
use Markdown::Pod;

our @EXPORT_OK = qw( get_pod markdown_to_pod );

sub get_pod {
    my $fh = shift;

    my $pod = do { local $/; <$fh> };
    $pod =~ s/\s+\Z//gsm;

    return $pod;
}

sub markdown_to_pod {
    my $file = shift;

    open my $fh, $file
        or warn("cannot open $file: $!\n"), return;
    my $markdown = do { local $/; <$fh> };
    close $fh;

    my $m2p = Markdown::Pod->new;
    my $pod = $m2p->markdown_to_pod(
        encoding => 'utf8',
        markdown => decode_utf8($markdown),
    );
    $pod =~ s/\s+\Z//gsm;

    return $pod;
}

1;


lib/Markdown/Pod.pm
package Markdown::Pod;
# ABSTRACT: Convert Markdown to POD

use strict;
use warnings;

our $VERSION = '0.009';

use Encode qw( encodings );
use List::Util qw( first );
use Markdent::Parser;
use Markdent::Types qw( Str );
use Markdown::Pod::Handler;
use Moose;
use MooseX::Params::Validate qw( validated_list );
use MooseX::StrictConstructor;
use namespace::autoclean;

sub markdown_to_pod {
    my $self = shift;
    my ( $dialect, $markdown, $encoding ) = validated_list(
        \@_,
        dialect  => { type => 'Str', default => 'Standard', optional => 1 },
        markdown => { type => 'Str' },
        encoding => { type => 'Str', default => q{},        optional => 1 },
    );

    my $capture = q{};
    open my $fh, '>', \$capture
        or die $!;

    if ($encoding) {
        my $found = first { $_ eq $encoding } Encode->encodings;
        if ($found) {
            binmode $fh, ":encoding($encoding)";
        }
        else {
            warn "cannot find such '$encoding' encoding\n";
        }
    }

    my $handler = Markdown::Pod::Handler->new(
        encoding => $encoding,
        output   => $fh,
    );

    my $parser = Markdent::Parser->new(
        dialect => $dialect,
        handler => $handler,
    );

    $parser->parse( markdown => $markdown );

    close $fh;

    $capture =~ s/\n+$/\n/;

    #
    # FIXME
    # dirty code to support blockquote
    #
    # UPDATE A.Speer - not needed, blockquote converted to use =over 2, =back
    #
    #$capture =~ s{
    #    ^ =begin \s+ blockquote
    #    \s+
    #    (.*?)
    #    \s+
    #    ^ =end \s+ blockquote
    #}{
    #    my $quoted = $1;
    #    $quoted =~ s/^/    /gsm;
    #    $quoted;
    #}xgsme;

    return $capture;
}

__PACKAGE__->meta->make_immutable;
1;
__END__


=head1 SYNOPSIS

    use Markdown::Pod;
    
    my $m2p = Markdown::Pod->new;
    my $pod = $m2p->markdown_to_pod(
        markdown => $markdown,
    );


=head1 DESCRIPTION

This module parses Markdown text and return POD text.
It uses L<Markdent> module to parse Markdown.
Due to POD doesn't support blockquoted HTML tag,
so quoted text of Markdown will not be handled properly.
Quoted text will be converted to POD verbatim section.


=attr markdown

markdown text

=attr encoding

encoding to use. Available type of encoding is same as L<Encode> module.

=method new

create Markdown::Pod object

=method markdown_to_pod

convert markdown text to POD text


=head1 SEE ALSO

=over

=item *

L<Markdent>

=item *

L<Pod::Markdown>

=item *

L<Text::MultiMarkdown>, L<Text::Markdown>

=back


=cut


lib/Markdown/Pod/script.pm
package Markdown::Pod::script;
# ABSTRACT: script to convert Markdown to POD

use strict;
use warnings;

our $VERSION = '0.009';

use Encode qw( encodings decode );
use Getopt::Long;
use List::Util qw( first );
use Markdown::Pod;

sub new {
    my $class = shift;

    bless {
        encoding => 'utf8',
        dialect  => 'Standard',
        verbose  => undef,
        argv     => [],
        @_,
    }, $class;
}

sub doit {
    my $self = shift;

    if ( my $action = $self->{action} ) {
        $self->$action() and return 1;
    }

    $self->show_help(1)
        unless @{ $self->{argv} } || $self->{load_from_stdin};

    if ( $self->{encoding} ) {
        my $found = first { $_ eq $self->{encoding} } Encode->encodings;
        if ( !$found ) {
            warn "cannot find such '$self->{encoding}' encoding\n";
            $self->show_help(1);
        }
    }

    my $m2p = Markdown::Pod->new;

    if ( $self->{load_from_stdin} ) {
        my $markdown = do { local $/; <STDIN>; };

        my $pod = $m2p->markdown_to_pod(
            encoding => $self->{encoding},
            dialect  => $self->{dialect},
            markdown => $markdown,
        );
        print $self->{encoding} ? decode( $self->{encoding}, $pod ) : $pod;
    }

    for my $file ( @{ $self->{argv} } ) {
        open my $fh, $file
            or warn("cannot open $file: $!\n"), next;
        my $markdown = do { local $/; <$fh>; };
        close $fh;

        my $pod = $m2p->markdown_to_pod(
            encoding => $self->{encoding},
            dialect  => $self->{dialect},
            markdown => $markdown,
        );
        print $self->{encoding} ? decode( $self->{encoding}, $pod ) : $pod;
    }

    return 1;
}

sub env {
    my ( $self, $key ) = @_;
    return $ENV{ "PERL_MARKDOWN2POD_" . $key } || q{};
}

sub parse_options {
    my $self = shift;

    local @ARGV = @{ $self->{argv} };
    push @ARGV, split /\s+/, $self->env('OPT');
    push @ARGV, @_;

    Getopt::Long::Configure("bundling");
    Getopt::Long::GetOptions(
        'e|encoding=s' => sub { $self->{encoding} = $_[1] },
        'd|dialect=s'  => sub { $self->{dialect}  = $_[1] },
        'v|verbose'    => sub { $self->{verbose}  = 1 },
        'h|help'       => sub { $self->{action}   = 'show_help' },
        'V|version'    => sub { $self->{action}   = 'show_version' },
    );

    if ( @ARGV == 0 || !-t STDIN ) {
        $self->{load_from_stdin} = 1;
    }

    $self->{argv} = \@ARGV;
}

sub show_help {
    my $self = shift;

    if ( $_[0] ) {
        die <<'USAGE';
Usage: markdown2pod [OPTIONS] <file1> [ <file2> ... ]

Try `markdown2pod --help` for details.
USAGE
    }

    print <<"HELP";
Usage: markdown2pod [OPTIONS] <file1> [ <file2> ... ]

OPTIONS:
    -e,--encoding       set markdown encoding
    -d,--dialect        set markdown dialect (Standard, Github, Theory)
    -v,--verbose        print more information
    -h,--help           print this message
    -V,--version        display software version

HELP

    return 1;
}

sub show_version {
    print "markdown2pod (Markdown::Pod) version $Markdown::Pod::VERSION\n";
    return 1;
}

1;
__END__

=head1 SYNOPSIS

    my $app = Markdown::Pod::script->new;
    $app->parse_options(@ARGV);
    $app->doit or exit(1);


=head1 DESCRIPTION

This module contains script related functions.


=method new

=method doit

=method env

=method parse_options

=method show_help

=method show_version

=cut


lib/Markdown/Pod/Handler.pm
package Markdown::Pod::Handler;
# ABSTRACT: Parser module to convert from markdown to POD

use strict;
use warnings;

our $VERSION = '0.009';

use Markdent::Types;

use namespace::autoclean;
use Moose;
use MooseX::SemiAffordanceAccessor;
use MooseX::Params::Validate qw( validated_list validated_hash );
use List::Util;

with 'Markdent::Role::EventsAsMethods';

has encoding => (
    is      => 'ro',
    isa     => t('Str'),
    default => q{},
);

has _output => (
    is       => 'ro',
    isa      => t('OutputStream'),
    required => 1,
    init_arg => 'output',
);

#  Default width for horizontal rule
#
our $HORIZONTAL_RULE_WIDTH = 80;

my $link_buf;
my $code_buf;
my $tble_buf;
my @tble = ( [] );
my @blockquotes;
my @list_type;

use constant {
    STACK_LINK     => 1,
    STACK_CODE     => 2,
    STACK_TBLE     => 3,
    STACK_STRONG   => 4,
    STACK_EMPHASIS => 5,
};

my @style_stack;

my $list_depth = 0;

sub _stream {
    my ( $self, @params ) = @_;
    print { $self->_output } @params;
}

sub start_document {
    my $self = shift;

    $self->_stream( '=encoding ' . $self->encoding . "\n\n" ) if $self->encoding;
}

sub end_document { }

sub text {
    my $self = shift;
    my ($text) = validated_list( \@_, text => { type => 'Str' } );

    if (@style_stack) {
        # This allows the end_link() handler to know that *some* text was inside
        # it. So if one has [`text`](http://example.org/), the end_code()
        # handler will output the code to the stream before the end_link()
        # finishes.
        $link_buf->{text} = '' if grep { $_ == STACK_LINK } @style_stack;

        if ( $style_stack[-1] == STACK_LINK ) {
            $link_buf->{text} = $text;
        }
        elsif ( $style_stack[-1] == STACK_CODE ) {
            $code_buf->{text} = $text;
        }
        elsif ( $style_stack[-1] == STACK_TBLE ) {
            $tble_buf->{text} = $text;
        }
        else {
            # another kind of style that does not require storing state
            $self->_stream($text);
        }
    }
    else {
        $self->_stream($text);
    }
}

sub start_header {
    my $self = shift;
    my ($level) = validated_list( \@_, level => { type => 'HeaderLevel' }, );

    $self->_stream("\n=head$level ");
}

sub end_header {
    my $self = shift;
    my ($level) = validated_list( \@_, level => { type => 'HeaderLevel' }, );

    $self->_stream("\n");
}

sub start_paragraph {
    my $self = shift;
}

sub end_paragraph {
    my $self = shift;

    $self->_stream("\n");
}

sub start_link {
    my $self = shift;
    my %p    = validated_hash(
        \@_,
        uri            => { type => 'Str' },
        title          => { type => 'Str', optional => 1 },
        id             => { type => 'Str', optional => 1 },
        is_implicit_id => { type => 'Bool', optional => 1 },
    );

    delete @p{ grep { !defined $p{$_} } keys %p };

    push @style_stack, STACK_LINK;
    $link_buf->{uri} = $p{uri};
    $self->_stream('L<');
}

sub end_link {
    my $self = shift;

    if ( $link_buf && exists $link_buf->{text} ) {
        $self->_stream("$link_buf->{text}|$link_buf->{uri}>");
    }
    else {
        $self->_stream("$link_buf->{uri}>");
    }

    pop @style_stack;
    $link_buf = undef;
}

sub start_strong {
    my $self = shift;

    push @style_stack, STACK_STRONG;
    $self->_stream('B<');
}

sub end_strong {
    my $self = shift;

    pop @style_stack;
    $self->_stream('>');
}

sub start_emphasis {
    my $self = shift;

    push @style_stack, STACK_EMPHASIS;
    $self->_stream('I<');
}

sub end_emphasis {
    my $self = shift;

    pop @style_stack;
    $self->_stream('>');
}

sub preformatted {
    my $self = shift;
    my ($text) = validated_list( \@_, text => { type => 'Str' }, );

    chomp $text;
    $text =~ s/^/    /gsm;
    $self->_stream( $text, "\n\n" );
}

sub start_blockquote {
    my $self = shift;

    $self->_stream("=over 2\n\n");
}

sub end_blockquote {
    my $self = shift;

    $self->_stream("=back\n\n");
}

sub start_unordered_list {
    my $self = shift;

    $self->_stream("\n") if $list_depth;
    $list_depth++;
    $self->_stream("=over\n\n");
}

sub end_unordered_list {
    my $self = shift;

    $list_depth--;
    $self->_stream("=back\n\n");
}

sub start_ordered_list {
    my $self = shift;

    $self->_stream("\n") if $list_depth;
    $list_depth++;
    $self->_stream("=over\n\n");
}

sub end_ordered_list {
    my $self = shift;

    $list_depth--;
    $self->_stream("=back\n\n");
}

sub start_list_item {
    my $self = shift;
    my %p = validated_hash( \@_, bullet => { type => 'Str' }, );

    $self->_stream("=item $p{bullet}\n\n");
}

sub end_list_item {
    my $self = shift;

    $self->_stream("\n\n");
}

sub start_code {
    my $self = shift;
    #  Start buffering this snippet
    push @style_stack, STACK_CODE;
    $code_buf = {};
}

sub end_code {
    my $self = shift;
    my $text = $code_buf->{'text'};
    if ( $text =~ /\n/m ) {
        #  Multi-line. Probably code block
        #
        $text =~ s/^(.*)$/ $1/mg;
        $self->_stream($text);
    }
    else {
        #  Single line
        #
        if ( $text =~ /[<>]/ ) {
            # this is so that extra angle brackets are not used unless necessary
            my @all_angle     = $text =~ /(<+|>+)/g;
            my @all_angle_len = map { length $_ } @all_angle;
            my $longest       = List::Util::max @all_angle_len;

            my $start_angle = "<" x ( $longest + 2 );
            my $end_angle   = ">" x ( $longest + 2 );
            $self->_stream("C$start_angle $text $end_angle");
        }
        else {
            $self->_stream("C<$text>");
        }
    }
    pop @style_stack;
    $code_buf = undef;
}

sub code_block {
    my $self = shift;
    my ($code) = validated_list(
        \@_,
        code     => { type => 'Str' },
        language => { type => 'Str', optional => 1 }
    );
    $code =~ s/^(.*)$/ $1/mg;
    $self->_stream("\n$code\n");
}

sub image {
    my $self = shift;
    my %p    = validated_hash(
        \@_,
        alt_text       => { type => 'Str' },
        uri            => { type => 'Str', optional => 1 },
        title          => { type => 'Str', optional => 1 },
        id             => { type => 'Str', optional => 1 },
        is_implicit_id => { type => 'Bool', optional => 1 },
    );

    delete @p{ grep { !defined $p{$_} } keys %p };

    my $alt_text = exists $p{alt_text} ? qq|alt="$p{alt_text}"| : q{};

    my $attr = exists $p{title} ? $p{title} : q{};
    my $attr_text = q{};
    while ( $attr =~ s/(\S+)="(.*?)"// ) {
        $attr_text .= qq{ $1="$2"};
    }
    while ( $attr =~ /(\S+)=(\S+)/g ) {
        $attr_text .= qq{ $1="$2"};
    }

    $self->_stream(qq|=for html <img src="$p{uri}" $alt_text$attr_text />|);
}

sub start_html_tag {
    my $self = shift;
    my ( $tag, $attributes ) = validated_list(
        \@_,
        tag        => { type => 'Str' },
        attributes => { type => 'HashRef' },
    );
}

sub end_html_tag {
    my $self = shift;
    my ( $tag, $attributes ) = validated_list( \@_, tag => { type => 'Str' }, );
}

sub html_tag {
    my $self = shift;
    my ( $tag, $attributes ) = validated_list(
        \@_,
        tag        => { type => 'Str' },
        attributes => { type => 'HashRef' },
    );

    my $attributes_str = q{};
    $attributes_str = join q{ },
        map {
            defined $attributes->{$_}
                ? qq|$_="$attributes->{$_}"|
                : qq|$_|
            } sort keys %$attributes;
    if ( $tag =~ /^br$/i ) {
        if ($attributes_str) {
            $self->_stream(qq|<$tag $attributes_str />\n|);
        }
        else {
            $self->_stream(qq|<$tag />\n|);
        }
    }
    else {
        if ($attributes_str) {
            $self->_stream(qq|<$tag $attributes_str />|);
        }
        else {
            $self->_stream(qq|<$tag />|);
        }
    }
}

sub html_block {
    my $self = shift;
    my ($html) = validated_list( \@_, html => { type => 'Str' }, );

    chomp $html;
    $self->_output()->print(
        <<"END_HTML"

=begin html

$html

=end html

END_HTML
    );
}

sub line_break {
    my $self = shift;
    $self->_stream("\n\n");
}

sub html_entity {
    my $self = shift;
    my ($entity) = validated_list( \@_, entity => { type => 'Str' } );

    $self->_stream("E<$entity>");
}

# Added A.Speer
sub horizontal_rule {
    my $self = shift;
    $self->_stream( ( '=' x $HORIZONTAL_RULE_WIDTH ) . "\n" );
}

sub auto_link {
    my $self = shift;
    my ($uri) = validated_list( \@_, uri => { type => 'Str' } );
    $self->_stream("L<$uri>");
}

sub html_comment_block {
    my $self = shift;
    # Stub
}

sub start_table {
    my $self = shift;
    # Stub
}

sub start_table_body {
    my $self = shift;
    # Stub
}

sub start_table_row {
    my $self = shift;
    # Stub
}

sub start_table_cell {
    my $self = shift;
    push @style_stack, STACK_TBLE;
    $tble_buf = {};
}

sub end_table {
    my $self = shift;
    eval {
        require Text::Table::Tiny;
        1;
    }
        || die('unable to load Text::Table::Tiny - please make sure it is installed !');
    my $table =
        Text::Table::Tiny::table( rows => \@tble, separate_rows => 0, header_row => 0 );
    #  Indent so table appears as POD code. Open to other suggestions
    $table =~ s/^(.*)/  $1/mg;
    $table .= "\n";
    #  Safety in case parser skips end-cell - which it seems to do sometimes
    pop @style_stack;
    $tble_buf = undef;
    $self->_stream($table);
}

sub end_table_body {
    my $self = shift;
    #  Safety
    pop @style_stack;
    $tble_buf = undef;
}

sub end_table_row {
    my $self = shift;
    push @tble, [];
    #  Safety
    pop @style_stack;
    $tble_buf = undef;
}

sub end_table_cell {
    my $self = shift;
    push @{ $tble[$#tble] }, $tble_buf->{'text'};
    #  Stop buffering table text
    pop @style_stack;
    $tble_buf = undef;
}

__PACKAGE__->meta->make_immutable;
no Moose;
1;
__END__

=for Pod::Coverage
    STACK_CODE
    STACK_EMPHASIS
    STACK_LINK
    STACK_STRONG
    STACK_TBLE
    auto_link
    code_block
    end_blockquote
    end_code
    end_document
    end_emphasis
    end_header
    end_html_tag
    end_link
    end_list_item
    end_ordered_list
    end_paragraph
    end_strong
    end_table
    end_table_body
    end_table_cell
    end_table_row
    end_unordered_list
    horizontal_rule
    html_block
    html_comment_block
    html_entity
    html_tag
    image
    line_break
    preformatted
    start_blockquote
    start_code
    start_document
    start_emphasis
    start_header
    start_html_tag
    start_link
    start_list_item
    start_ordered_list
    start_paragraph
    start_strong
    start_table
    start_table_body
    start_table_cell
    start_table_row
    start_unordered_list
    text

=head1 SYNOPSIS

    my $handler = Markdown::Pod::Handler->new(
        encoding => $encoding,
        output   => $fh,
    );
    
    my $parser = Markdent::Parser->new(
        dialect => $dialect,
        handler => $handler,
    );



=head1 DESCRIPTION

This module is a handler of L<Markdent> Markdown parser.
It converts Markdown to POD.


=attr markdown

markdown text

=attr encoding

encoding to use


=method new

create Markdown::Pod::Handler object

=method markdown_to_pod

convert markdown text to POD text


=head1 SEE ALSO

=over

=item *

L<Markdent>

=item *

L<Pod::Markdown>

=item *

L<Text::MultiMarkdown>, L<Text::Markdown>

=back


=cut


Makefile.PL
use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.17;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Convert Markdown to POD",
  "AUTHOR" => "\x{ae40}\x{b3c4}\x{d615} - Keedi Kim <keedi\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "Markdown-Pod",
  "EXE_FILES" => [
    "bin/markdown2pod"
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010",
  "NAME" => "Markdown::Pod",
  "PREREQ_PM" => {
    "Getopt::Long" => 0,
    "List::Util" => 0,
    "Markdent" => "0.40",
    "Moose" => 0,
    "MooseX::Params::Validate" => 0,
    "MooseX::SemiAffordanceAccessor" => 0,
    "MooseX::StrictConstructor" => 0,
    "Text::Table::Tiny" => 0,
    "constant" => 0,
    "namespace::autoclean" => 0
  },
  "TEST_REQUIRES" => {
    "Exporter" => "5.57",
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.009",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Exporter" => "5.57",
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Getopt::Long" => 0,
  "List::Util" => 0,
  "Markdent" => "0.40",
  "Moose" => 0,
  "MooseX::Params::Validate" => 0,
  "MooseX::SemiAffordanceAccessor" => 0,
  "MooseX::StrictConstructor" => 0,
  "Test::More" => 0,
  "Text::Table::Tiny" => 0,
  "constant" => 0,
  "namespace::autoclean" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
